blog/_src/posts/2016-11-02-beta-reduction-part-1.md

```diff
@@ -0,0 +1,189 @@
+
+# Some Examples
+The first example evaluates the same way in all of the languages.
+
+```
+    (λx.x) (λy.y)
+->β x[ x := (λy.y) ]
+=   (λy.y)
+```
+
+```racket
+    ((λ (x) x) (λ (y) y))
+->β x[ x := (λ (y) y) ]
+=   (λ (y) y)
+```
+
+```ocaml
+    (fun x -> x) (fun y -> y)
+->β x[x := (fun y -> y) ]
+=   x[x := (fun y -> y) ]
+```
+
+```haskell
+    (\x -> x) (\y -> y)
+->β x[ x := (\y -> y) ]
+=   (\y -> y)
+```
+
+In our next example, we will see the differences between the languages.  They all reduce to the same thing, albeit in different ways.
+
+```haskell
+    (\x -> \y -> y) ((\z -> z) (\a -> a))
+->β (\y -> y)[x := ((\z -> z) (\a -> a))]
+= (\y -> y)
+```
+
+Note that the application on the right hand side is never evaluated.  In an eager language like Racket or OCaml, the term being substituted must be a value, so the evaluation follows a different path.
+
+```racket
+    ((λ (x) (λ (y) y)) ((λ (z) z) (λ (a) a)))
+->β ((λ (x) (λ (y) y)) (z[ z := (λ (a) a) ]))
+=   ((λ (x) (λ (y) y)) (λ (a) a))
+->β (λ (y) y)[ x := (λ (a) a) ]
+=   (λ (y) y)
+```
+
+```ocaml
+    (fun x -> (fun y -> y)) ((fun z -> z) (fun a -> a))
+->β (fun x -> (fun y -> y)) (z[ z := (fun a -> a) ])
+=   (fun x -> (fun y -> y)) (fun a -> a)
+->β (fun y -> y)[ x := (fun a -> a) ]
+=   (fun y -> y)
+```
+
+The nondeterministic semantics of the λ-calculus lead to two possible paths through the above computation:
+
+```
+    (λx.λy.y) ((λz.z) (λa.a))
+->β (λx.λy.y) (z[ z := (λa.a) ])
+=   (λx.λy.y) (λa.a)
+->β (λy.y)[ x := (λa.a) ]
+=   (λy.y)
+```
+
+```
+    (λx.λy.y) ((λz.z) (λa.a))
+->β (λy.y)[ x := ((λz.z) (λa.a)) ]
+=   (λy.y)
+```
+
+Lets look at a final example.  This one is more complicated than the previous ones.  I'm also going to stop showing the substitution explicitly.  In the literature, it is fairly common not to see it explicitly, but you should still be able to follow what's going on.
+
+```racket
+    ((lambda (x) x) (((lambda (a) (lambda (b) (a b))) (lambda (y) y)) (lambda (z) z)))
+->β ((lambda (x) x) ((lambda (b) ((lambda (y) y) b)) (lambda (z) z)))
+->β ((lambda (x) x) ((lambda (y) y) (lambda (z) z)))
+->β ((lambda (x) x) (lambda (z) z))
+->β (lambda (z) z)
+```
+
+The same thing happens in OCaml
+
+```ocaml
+    (fun x -> x) ((fun a -> (fun b -> a b)) (fun y -> y) (fun z -> z));;
+->β (fun x -> x) ((fun b -> (fun y -> y) b) (fun z -> z));;
+->β (fun x -> x) ((fun y -> y) (fun z -> z))
+->β (fun x -> x) (fun z -> z)
+->β (fun z -> z)
+```
+
+In Haskell, the situation is a little different:
+
+```haskell
+    (\x -> x) ((\a -> \b -> a b) (\y -> y) (\z -> z))
+->β (\a -> \b -> a b) (\y -> y) (\z -> z)
+->β (\b ->  (\y -> y) b) (\z -> z)
+->β (\y -> y) (\z -> z)
+->β (\z -> z)
+```
```

Comment:
Again this is fine. One can only see the difference between call-by-name and call-by-need in function bodies where the variable occurs more than once. Your examples are all affine lambda-terms (in the sense that they can be typed in an affine type system, where variables may be erased but not duplicated), so there is no difference.